flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -819,7 +819,7 @@ public void shutdownFetcher() {
                 LOG.warn("Encountered exception closing record publisher factory", e);
             }
         } finally {
-            shardConsumersExecutor.shutdownNow();
+            shardConsumersExecutor.shutdown();
```

Comment:
Are you sure this is correct? `shutdown()` [will not interrupt active shard consumers](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html#shutdown()). How will the running threads get interrupted?

Reply:
The whole point of https://issues.apache.org/jira/browse/FLINK-23528 is to get rid of interruptions and thus avoid inconsistent state during stop-with-savepoint. 
Instead of interruptions, the poll loop should simply return by checking the `running` flag. That means of course that shutdown may take as long as the timeout of a specific poll request. Can we tweak that? Afaik it boils down to how long `com.amazonaws.services.kinesis.AmazonKinesis#getRecords` blocks.

Reply:
For the `POLLING` record consumer you are correct. But for the `FanOutRecordPublisher` (EFO) it uses a reactive stream and does not check the `running` flag until subscription is complete/errors (up to 5 minutes). That being said, the `closeRecordPublisherFactory()` method will kill the Netty threads and close the client which should cause the `running` flag to be checked. We would need to check for edge cases

Reply:
I'll try to add a test case for that and then we'll see.

Reply:
There are no integration tests for EFO since Kinesalite does not support EFO. Coincidentally I was recently discussing adding e2e test for EFO using AWS creds. We need to work out how to supply AWS creds to make this happen